### PR TITLE
Keep the encoding of standard output & error consistent with the console code page

### PR DESCRIPTION
**Pull request by GangWang01**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1388756/

### Context
With OS language English and VS language JPN, building newly created CMake project named `プロジェクト` got garbage characters in the output.

### Changes Made
Make the encoding of process standard output & error consistent with the console code page.

### Testing
Added unit tests.

### CODE REVIEW

**Review by rokonec — APPROVED** [review]

**Comment by Forgind** [comment]
This doesn't seem like a bad change, and I haven't checked whether it actually fixes the bug or not. It looks like I looked at this bug over two years ago and thought the problem may have been because the Message task didn't support changing the code page directly and may have been that the user needed to change how they were using a ToolTask. Assuming this does fix the problem, though this sounds like a positive change, I'd say it suggests the problem should have been fixed by whoever implemented the ToolTask that is ultimately broken; all they would've had to do is set that property as you did here except in their ToolTask extension. In fact, I'd argue that's a better fix, as this one changes fairly fundamental behavior in MSBuild, which always has potential to cause someone issues even if the change looks benign.

Just my 2c 🙂

File: src/Utilities/ToolTask.cs [hunk 1]
```diff
@@ -1404,6 +1404,9 @@ public override bool Execute()
                         }
 
                         File.AppendAllText(_temporaryBatchFile, commandLineCommands, encoding);
+                        // Keep the encoding of standard output & error consistent with the console code page.
+                        StandardOutputEncoding = encoding;
+                        StandardErrorEncoding = encoding;
```
Comment (JanKrivanek) [inline src/Utilities/ToolTask.cs:1409]: Looks good - it just might be sfer to put this behind changewave (in case it breaks anybody)

**Review by JanKrivanek — APPROVED** [review]
Overall looks good

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Please try to see if we can can move the logic to property initialization.

If not - let's discuss other alternatives

File: src/Utilities/ToolTask.cs [hunk 2]
```diff
@@ -1417,6 +1417,12 @@ public override bool Execute()
                         }
 
                         File.AppendAllText(_temporaryBatchFile, commandLineCommands, encoding);
+                        if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10))
+                        {
+                            // Keep the encoding of standard output & error consistent with the console code page.
+                            StandardOutputEncoding = encoding;
+                            StandardErrorEncoding = encoding;
+                        }
```
[the remark this answers is not in the document]
Comment (GangWang01) [reply src/Utilities/ToolTask.cs:1456]: Agree.

**Review by JanKrivanek — APPROVED** [review]
Looks good!

**Pull request merged**